src/routes/users.rs

```diff
@@ -1,35 +1,73 @@
+use anyhow::Context;
 use askama_axum::IntoResponse;
 use axum::{
-    extract::{Query, State},
-    response::{Redirect, Response},
-    routing::{get, post},
-    Router,
+    extract::{Query, State}, response::{Redirect, Response}, routing::{get, post}, Router
 };
+use tokio::task;
 use garde::{Report, Validate};
-use serde::Deserialize;
 use tower_sessions::Session;
 
 use crate::{
-    authentication::{self, AuthUser},
-    extract::{self, qs_form::QsForm},
-    forms::users::Login,
-    response_error::ResponseResult,
-    server::AppState,
-    views::{
+    authentication::{self, AuthUser}, extract::{self, qs_form::QsForm}, forms::users::{Login, CreateOAuthUser}, response_error::ResponseResult, server::AppState, views::{
         layout::LayoutTemplate,
         login::{DemoLoginTemplate, LoginTemplate},
         users::ProfileTemplate,
-    },
+    }
+};
+use serde::{Deserialize, Serialize};
+
+use openidconnect::core::{
+    CoreAuthDisplay, CoreClaimName, CoreClaimType, CoreClient, CoreClientAuthMethod, CoreGrantType,
+    CoreIdTokenVerifier, CoreJsonWebKey, CoreJsonWebKeyType, CoreJsonWebKeyUse,
+    CoreJweContentEncryptionAlgorithm, CoreJweKeyManagementAlgorithm, CoreJwsSigningAlgorithm,
+    CoreResponseMode, CoreResponseType, CoreRevocableToken, CoreSubjectIdentifierType,
+};
+use openidconnect::reqwest::http_client;
+use openidconnect::{
+    AdditionalProviderMetadata, AuthenticationFlow, AuthorizationCode, ClientId, ClientSecret,
+    CsrfToken, IssuerUrl, Nonce, OAuth2TokenResponse, ProviderMetadata, RedirectUrl, RevocationUrl,
+    Scope,
 };
 
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct RevocationEndpointProviderMetadata {
+    revocation_endpoint: String,
+}
+impl AdditionalProviderMetadata for RevocationEndpointProviderMetadata {}
+type GoogleProviderMetadata = ProviderMetadata<
```

Comment:
I think this type, and the `OidcSession` struct would be better off in a separate `oauth.rs`, `oidc.rs` or `sso.rs` file.